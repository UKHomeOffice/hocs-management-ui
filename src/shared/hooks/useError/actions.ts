import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
