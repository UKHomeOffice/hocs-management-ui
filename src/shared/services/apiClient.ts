import axios from 'axios';

export default {
    createClient: () => {
        const client = axios.create();

        return {
            get: client.get,
            delete: client.delete,
            post: client.post,
            put: client.put
        };
    }
};
