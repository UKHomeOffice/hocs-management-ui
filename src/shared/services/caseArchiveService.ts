import axios from 'axios';
import ArchiveCaseModel from '../models/archiveCaseModel';

export const archiveCase = (archiveCaseModel: ArchiveCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/archive', archiveCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
