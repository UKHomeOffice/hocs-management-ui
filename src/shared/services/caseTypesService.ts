import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
