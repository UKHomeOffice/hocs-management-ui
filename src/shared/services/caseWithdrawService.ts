import axios from 'axios';
import WithdrawCaseModel from '../models/withdrawCaseModel';

export const withdrawCase = (withdrawCaseModel: WithdrawCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/withdraw', withdrawCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
