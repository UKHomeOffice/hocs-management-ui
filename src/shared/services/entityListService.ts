import axios from 'axios';
import EntityListItem from 'shared/models/entityListItem';

export const getListItems = (listName: string) => new Promise<EntityListItem[]>((resolve, reject) => axios
    .get(`/api/entity/list/${listName}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createListItem = (item: EntityListItem, listName: string) => new Promise((resolve, reject) => axios
    .post(`/api/entity/list/${listName}`, item)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);

export const getItemDetails = (itemUUID: string) => new Promise<EntityListItem>((resolve, reject) => axios
    .get(`/api/entity/${itemUUID}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const updateListItem = (item: EntityListItem, listName: string) => new Promise((resolve, reject) => axios
    .post(`/api/entity/list/update/${listName}`, item)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);

export const deleteListItem = (uuid: string, listName: string) => new Promise((resolve, reject) => axios
    .delete(`/api/entity/list/${listName}`, { data: { entityUUID: uuid } })
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
