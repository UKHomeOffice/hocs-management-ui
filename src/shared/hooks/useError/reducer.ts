import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
