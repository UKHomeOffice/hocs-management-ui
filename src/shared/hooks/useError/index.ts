import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
