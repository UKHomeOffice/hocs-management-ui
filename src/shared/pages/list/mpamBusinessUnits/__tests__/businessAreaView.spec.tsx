import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, wait, render, RenderResult } from '@testing-library/react';
import BusinessAreaView from '../businessAreaView';
import * as ListService from '../../../../services/entityListService';
import * as useError from '../../../../hooks/useError';

let match: match<any>;
let history: History<any>;
let location: Location;

jest.mock('../../../../services/entityListService', () => ({
    __esModule: true,
    getListItems: jest.fn().mockReturnValue(Promise.resolve({
        data: [{
            simpleName: 'testSimpleName1',
            uuid: 'testId1',
            title: 'testTitle1'
        }, {
            simpleName: 'testSimpleName2',
            uuid: 'testId2',
            title: 'testTitle2'
        }]
    }))
}));

const getListItemsSpy = jest.spyOn(ListService, 'getListItems');
const useErrorSpy = jest.spyOn(useError, 'default');
const setMessageSpy = jest.fn();
const clearErrorsSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <BusinessAreaView history={history} location={location} match={match}></BusinessAreaView>
    </MemoryRouter>
);

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: {},
        path: '',
        url: ''
    };
    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    useErrorSpy.mockImplementation(() => [{}, jest.fn(), clearErrorsSpy, setMessageSpy]);
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
});

describe('when the businessAreaView component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(2);
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(getListItemsSpy).toHaveBeenCalled();
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});
