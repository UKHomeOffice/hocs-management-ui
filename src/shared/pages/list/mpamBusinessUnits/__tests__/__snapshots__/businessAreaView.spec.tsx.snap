// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the businessAreaView component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <a
      class="govuk-back-link"
      href="/select-business-area"
    >
      Back
    </a>
    <div>
      <h1
        class="govuk-heading-xl"
      >
        View and Edit Business Units for 
      </h1>
      <div>
        <table
          class="govuk-table"
        >
          <thead
            class="govuk-table__head"
          >
            <tr
              class="govuk-table__row"
            >
              <th
                class="govuk-table__header"
                scope="col"
              >
                Business Unit
              </th>
              <th
                class="govuk-table__header"
                scope="col"
              >
                Action
              </th>
            </tr>
          </thead>
          <tbody
            class="govuk-table__body"
          >
            <tr
              class="govuk-table__row"
            >
              <td
                class="govuk-table__cell"
              >
                testTitle1
              </td>
              <td
                class="govuk-table__cell"
              >
                <a
                  href="#"
                >
                  Amend
                </a>
              </td>
            </tr>
            <tr
              class="govuk-table__row"
            >
              <td
                class="govuk-table__cell"
              >
                testTitle2
              </td>
              <td
                class="govuk-table__cell"
              >
                <a
                  href="#"
                >
                  Amend
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <button
        class="govuk-button govuk-!-margin-right-1 add-team-members-button"
        data-module="govuk-button"
        type="submit"
      >
        Add Business Unit
      </button>
    </div>
  </div>
</div>
`;
